### PR TITLE
Implementation of Update Cart endpoint - Step2

### PR DESCRIPTION
**Pull request by dumega**
# Description
Adds support for adding/removing coupon

Prev PR https://github.com/BoltApp/bolt-magento2/pull/853
Next PR https://github.com/BoltApp/bolt-magento2/pull/856

Fixes: (link Jira ticket)

#changelog Implementation of Update Cart endpoint - Step3

# Type of change

- [ ] Bug fix (change which fixes an issue)
- [ ] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
Please validate that you have tested your change in at least one of the following areas:

- [ ] Successfully tested locally (or docker image)
- [ ] Successfully tested on a staging or sandbox server
- [ ] Successfully tested on a merchant's staging server

# For PR Reviewer 
- [ ] Reviewed unit tests to make sure we are using real components rather than mocks as much as possible?
- [ ] For any major change (observer, new Bolt feature, core Magento interaction) we must add a feature switch, did you verify this?

# Checklist:

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have created or modified unit tests to sufficiently cover my changes.
- [ ] I have added my Jira ticket link and provided a changelog message.

### CODE REVIEW

File: Helper/Cart.php [hunk 1]
```diff
@@ -1719,15 +1719,15 @@ public function collectDiscounts(
         // selecting specific shipping option, so the conditional statement should also
         // check if getCouponCode is not null
         /////////////////////////////////////////////////////////////////////////////////
-        if (( $amount = abs($address->getDiscountAmount()) ) || $address->getCouponCode()) {
+        if (( $amount = abs($address->getDiscountAmount()) ) || $quote->getCouponCode()) {
```
Comment (vitaliyreznikov) [inline Helper/Cart.php:1722]: Why we need this change?

File: Model/Api/UpdateCart.php [hunk 2]
```diff
@@ -121,8 +125,56 @@ public function execute($cart, $add_items = null, $remove_items = null, $discoun
                 $this->setShipment($cart['shipments'][0], $immutableQuote);
                 $this->setShipment($cart['shipments'][0], $parentQuote);
             }
+            // TODO : add/remove giftcard
```
Comment (vitaliyreznikov) [inline Model/Api/UpdateCart.php:128]: Please add TODO for clearing bolt order cache, we can implement it once finish the discussion about caching.

File: Helper/Cart.php (hunk 1 above)
Comment (dumega) [reply Helper/Cart.php:1749]: Cause $address->getCouponCode() could returns empty code in some contexts, and if we look into the core code of Magento, it usually use $address->getQuote()->getCouponCode(); to get coupon code. Correct me if I miss anything

File: Model/Api/UpdateCart.php (hunk 2 above)
Comment (dumega) [reply Model/Api/UpdateCart.php]: Done

**Review by vitaliyreznikov — APPROVED** [review]
LGTM

**Comment by vitaliyreznikov** [comment]
Tests in php 72 are really passed, issue with memory happens when we render results. Need to investigate, but it doesn't block the PR

**Pull request merged**